src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```diff
@@ -617,17 +631,60 @@ internal override bool IsRecord
         }
 
 #nullable enable
-        internal sealed override bool IsExtension => false; // PROTOTYPE Extension symbols not yet supported in metadata
+        internal sealed override bool IsExtension => TypeKind == TypeKind.Extension;
+
+        internal sealed override TypeSymbol? ExtensionUnderlyingTypeNoUseSiteDiagnostics
+        {
+            get
+            {
+                return GetDeclaredExtensionUnderlyingType();
+            }
+        }
 
-        internal sealed override TypeSymbol? ExtensionUnderlyingTypeNoUseSiteDiagnostics => null; // PROTOTYPE
         internal sealed override ImmutableArray<NamedTypeSymbol> BaseExtensionsNoUseSiteDiagnostics
-            => ImmutableArray<NamedTypeSymbol>.Empty; // PROTOTYPE
+        {
+            get
+            {
+                var declaredBaseExtensions = GetDeclaredBaseExtensions();
+                return declaredBaseExtensions.SelectAsArray(t => BaseTypeAnalysis.TypeDependsOn(depends: t, on: this) ? CyclicInheritanceError(t) : t);
+            }
+        }
 
         internal sealed override TypeSymbol? GetDeclaredExtensionUnderlyingType()
-            => throw ExceptionUtilities.Unreachable(); // PROTOTYPE
+        {
+            if (this.TypeKind != TypeKind.Extension)
+                return null;
+
+            var uncommon = GetUncommonProperties();
+            if (uncommon == s_noUncommonProperties)
+                return null;
+
+            if (uncommon.lazyDeclaredExtensionUnderlyingType is not null)
+                return uncommon.lazyDeclaredExtensionUnderlyingType;
+
+            DecodeExtensionType(out TypeSymbol underlyingType, out _);
```

Comment:
>out _ [](http://example.com/codeflow?start=63&length=5)

Not sure why we are ignoring result of some work. What prevents us from caching it?